daemon/api.go

```diff
@@ -967,6 +976,119 @@ func deleteCapability(c *Command, r *http.Request) Response {
 	}
 }
 
+// skillGrant is a part of response from getSkills().
+type skillGrant struct {
+	Snap string `json:"snap"`
+	Name string `json:"name"`
+}
+
+// skillInfo is a part of response from getSkills().
+type skillInfo struct {
+	Snap  string       `json:"snap"`
+	Name  string       `json:"name"`
+	Type  string       `json:"type"`
+	Label string       `json:"label"`
+	Slots []skillGrant `json:"slots"`
+}
+
+// getSkills returns a list of all the skills and which slots use them.
+func getSkills(c *Command, r *http.Request) Response {
+	var skills []skillInfo
+	for _, skill := range c.d.skills.AllSkills("") {
+		var slots []skillGrant
+		for _, slot := range c.d.skills.GrantsOf(skill.Snap, skill.Name) {
+			slots = append(slots, skillGrant{
+				Snap: slot.Snap,
+				Name: slot.Name,
+			})
+		}
+		skills = append(skills, skillInfo{
+			Snap:  skill.Snap,
+			Name:  skill.Name,
+			Type:  skill.Type,
+			Label: skill.Label,
+			Slots: slots,
+		})
+	}
+	return SyncResponse(skills)
+}
+
+// SkillAction is an action performed on the skill system.
+type SkillAction struct {
+	Action string       `json:"action"`
+	Skill  skills.Skill `json:"skill,omitempty"`
+	Slot   skills.Slot  `json:"slot,omitempty"`
+}
+
+// changeSkills controls the skill system.
+// Skills can be granted to and revoked from slots.
+// When enableInternalSkillActions is true skills and slots can also be
+// explicitly added and removed.
+func changeSkills(c *Command, r *http.Request) Response {
+	var a SkillAction
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&a); err != nil || a.Action == "" {
+		return BadRequest("cannot decode request body into a skill action: %v", err)
+	}
+	switch a.Action {
+	case "grant":
+		err := c.d.skills.Grant(a.Skill.Snap, a.Skill.Name, a.Slot.Snap, a.Slot.Name)
+		if err != nil {
+			return BadRequest("%v", err)
+		}
+		return SyncResponse(nil)
+	case "revoke":
+		err := c.d.skills.Revoke(a.Skill.Snap, a.Skill.Name, a.Slot.Snap, a.Slot.Name)
+		if err != nil {
+			return BadRequest("%v", err)
+		}
+		return SyncResponse(nil)
+	case "add-skill":
+		if !c.d.enableInternalSkillActions {
```

Comment:
I'm not quite sure about whether we want to block in such a way or to conditionally enable this on development mode. Either way, while we have that it would be better check these at once before entering the main switch, and safer to use whitelisting rather than blacklisting:
```
if !c.d.enableInternalSkillActions && a.Action != "grant" && a.Action != "revoke" {
        return BadRequest("...")
}
```

With this it will be harder for us to screw it up badly on future changes by mistake.